src/Categories/Double/Core.agda

```diff
@@ -0,0 +1,315 @@
+import Relation.Binary.Reasoning.Setoid as SetoidR
+
+import Categories.Category.Unbundled as Cat
+
+{-
+When defining a double category over a setoid enriched category
+we wish to state an identity law like the following:
+```
+
+          f
+      T₁  →  T₂
+    v ↓   s  ↓ w
+      B₁  →  B₂              f
+          g                T₁ → T₂
+          ∘        ≃    v  ↓    ↓ w
+          h                B₁ → B₂
+      B₁  →  B₂             h ∘ g
+   id ↓  id  ↓ id
+      B₁  →  B₂
+          h
+```
+
+Note that if we try to define homogeneous equality between squares of
+the form: `Sq₂ f (h ∘ g) v w`, then this will not work as the left-hand
+verical composite is a square of type `Sq₂ f (h ∘ g) (id ∘ v) (id ∘ w)`.
+Instead, we define equality of squares to mean an equalities in each of
+the horizontal or vertical hom setoids.
+
+This equality of squares in a double category is captured in the
+definition |SqEqualtiy|.
+
+-}
+module _ {o ℓ ℓ' e e' : Level} {Obj : Set o}
+         (Hor : Cat.Category Obj ℓ e) (Ver : Cat.Category Obj ℓ' e') where
+  private
+    module H = Cat.Category Hor
+    module V = Cat.Category Ver
+    _≈ₕ_ : ∀ {A B} → Rel (A H.⇒ B) e
+    _≈ₕ_ = H._≈_
+    _≈ᵥ_ : ∀ {A B} → Rel (A V.⇒ B) e'
+    _≈ᵥ_ = V._≈_
+  record SqEquality
+    {T₁ T₂ B₁ B₂ : Obj}
+    (hT₁ hT₂ : T₁ H.⇒ T₂)
+    (hB₁ hB₂ : B₁ H.⇒ B₂)
+    (vL₁ vL₂ : T₁ V.⇒ B₁)
+    (vR₁ vR₂ : T₂ V.⇒ B₂) : Set (e ⊔ e') where
+      field
+        horT≈ : hT₁ ≈ₕ hT₂
+        horB≈ : hB₁ ≈ₕ hB₂
+        verL≈ : vL₁ ≈ᵥ vL₂
+        verR≈ : vR₁ ≈ᵥ vR₂
+
+module _ {o ℓ ℓ' e e' : Level} {Obj : Set o}
+         (Hor : Cat.Category Obj ℓ e) (Ver : Cat.Category Obj ℓ' e') where
+  private
+    module H = Cat.Category Hor
+    module V = Cat.Category Ver
+  dual≈ :
+    ∀ {T₁ T₂ B₁ B₂ : Obj} →
+    {hT₁ hT₂ : T₁ H.⇒ T₂} →
+    {hB₁ hB₂ : B₁ H.⇒ B₂} →
+    {vL₁ vL₂ : T₁ V.⇒ B₁} →
+    {vR₁ vR₂ : T₂ V.⇒ B₂} →
+    SqEquality Hor Ver hT₁ hT₂ hB₁ hB₂ vL₁ vL₂ vR₁ vR₂ →
+    SqEquality Ver Hor vL₁ vL₂ vR₁ vR₂ hT₁ hT₂ hB₁ hB₂
+  dual≈ S = record
+      { horT≈ = S.verL≈
+      ; horB≈ = S.verR≈
+      ; verL≈ = S.horT≈
+      ; verR≈ = S.horB≈
+      }
+    where module S = SqEquality S
+
+-- Basic definition of a strict setoid-enriched |Double Category|.
+record Category (o ℓ ℓ' e e' : Level) : Set (suc (o ⊔ ℓ ⊔ e ⊔ ℓ' ⊔ e')) where
+  field
+    Obj : Set o
+    Hor : Cat.Category Obj ℓ e
+    Ver : Cat.Category Obj ℓ e
+  private
+    module H = Cat.Category Hor
+    module V = Cat.Category Ver
+  _⇒ₕ_ : Rel Obj ℓ
+  _⇒ₕ_ = H._⇒_
+  _≈ₕ_ : ∀ {A B} → Rel (A ⇒ₕ B) e
+  _≈ₕ_ = H._≈_
+  _⇒ᵥ_ : Rel Obj ℓ
+  _⇒ᵥ_ = V._⇒_
+  _≈ᵥ_ : ∀ {A B} → Rel (A ⇒ᵥ B) e
+  _≈ᵥ_ = V._≈_
+  _∘₁ₕ_ : ∀ {A B C} → (B ⇒ₕ C) → (A ⇒ₕ B) → (A ⇒ₕ C)
+  _∘₁ₕ_ = H._∘_
+  _∘₁ᵥ_ : ∀ {A B C} → (B ⇒ᵥ C) → (A ⇒ᵥ B) → (A ⇒ᵥ C)
+  _∘₁ᵥ_ = V._∘_
+
+  field
+    Sq₂ : {T₁ T₂ B₁ B₂ : Obj} → T₁ ⇒ₕ T₂ → B₁ ⇒ₕ B₂ → T₁ ⇒ᵥ B₁ → T₂ ⇒ᵥ B₂ → Set ℓ
+
+  Sq≈ : ∀ {T₁ T₂ B₁ B₂ : Obj}
```

Comment:
So this definition is odd: it reads like it's a relation between two squares, but it's not: it's a relation between two **perimeters**. The insides of the square are ignored!

This is a useful notion, but it should have a different name.

Reply:
And then actual 'square equality' should be defined to say that given two squares whose perimeters are equal, we get a type of witnesses that the squares themselves are equal.

Reply:
We could call them "frames" rather than perimeters to follow Mike Shulman's "framed Bicategory" terminology.

Reply:
If that's a correct use of the term (I don't know - reference?), then I definitely like it!